Source/WebKit/WebProcess/WebPage/WebPage.cpp

```diff
@@ -9414,6 +9414,32 @@ void WebPage::createTextFragmentDirectiveFromSelection(CompletionHandler<void(UR
     completionHandler(WTFMove(url));
 }
 
+void WebPage::getTextFragmentRanges(CompletionHandler<void(const Vector<EditingRange>&&)>&& completionHandler)
+{
+    RefPtr focusedOrMainFrame = protectedCorePage()->checkedFocusController()->focusedOrMainFrame();
+    if (!focusedOrMainFrame) {
+        completionHandler({ });
+        return;
+    }
+    RefPtr document = focusedOrMainFrame->document();
+
+    RefPtr frame = document->frame();
+    if (!frame) {
+        completionHandler({ });
+        return;
+    }
+
+    Vector<EditingRange> editingRanges;
+    if (RefPtr highlightRegistry = document->fragmentHighlightRegistryIfExists()) {
+        for (auto& highlight : highlightRegistry->map()) {
+            for (auto& highlightRange : highlight.value->highlightRanges())
+                editingRanges.append(EditingRange::fromRange(*frame, makeSimpleRange(highlightRange->range())));
```

Comment:
Something like:

```suggestion
                editingRanges.append(EditingRange::fromRange(*frame, makeSimpleRange(Ref { highlightRange->range() }.get())));
```

...or you could add a `protectedRange()` getter to `HighlightRange` that returns a `Ref<AbstractRange>`